File: java-microservice/withdrawal-service/src/main/java/com/bankapp/withdrawal_service/utils/TransactionLogger.java
package com.bankapp.withdrawal_service.utils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;

public class TransactionLogger {
    private static final Logger logger = LoggerFactory.getLogger(TransactionLogger.class);
    public static void log(String userId, String operationType, String accountId, String amount, String currency, String status, String message, String clientIp) {
        String logEntry = String.format("TIMESTAMP: %s | USER_ID: %s | OPERATION: %s | ACCOUNT_ID: %s | AMOUNT: %s | CURRENCY: %s | STATUS: %s | IP: %s | MESSAGE: %s",
                Instant.now().toString(),
                userId == null ? "N/A" : userId,
                operationType,
                accountId == null ? "N/A" : accountId,
                amount == null ? "N/A" : amount,
                currency == null ? "N/A" : currency,
                status,
                clientIp == null ? "N/A" : clientIp,
                message);
        logger.info("[TRANSACTION_TRACE] {}", logEntry);
    }
}

File: node-microservice/deposit-service/src/utils/transactionLogger.js
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const logFilePath = path.join(__dirname, '../../transaction_trace.txt');

const logOperation = async (logDetails) => {
  const timestamp = new Date().toISOString();
  const {
    userId = 'N/A',
    operationType,
    accountId = 'N/A',
    status,
    message = '',
    clientIp = 'N/A',
    amount = 'N/A',
    currency = 'N/A',
  } = logDetails;

  const logEntry = `TIMESTAMP: ${timestamp} | USER_ID: ${userId} | OPERATION: ${operationType} | ACCOUNT_ID: ${accountId} | AMOUNT: ${amount} | CURRENCY: ${currency} | STATUS: ${status} | IP: ${clientIp} | MESSAGE: ${message}\n`;

  try {
    await fs.appendFile(logFilePath, logEntry);
  } catch (error) {
    console.error('Error al escribir en el log de transacciones:', error);
  }
};

export { logOperation };
